corgidrp/calibrate_nonlin.py

```diff
@@ -0,0 +1,960 @@
+    check.real_positive_scalar(min_write, 'min_write', TypeError)
+    check.real_positive_scalar(max_write, 'max_write', TypeError)
+    if min_write >= max_write:
+        raise CalNonlinException('max_write must be greater than min_write')
+    if (norm_val < min_write) or (norm_val > max_write):
+        raise CalNonlinException('norm_val must be between min_write and '
+                                 'max_write.')
+    check.real_nonnegative_scalar(rms_low_limit, 'rms_low_limit', TypeError)
+    check.real_nonnegative_scalar(rms_upp_limit, 'rms_upp_limit', TypeError)
+    if rms_low_limit >= rms_upp_limit:
+        raise CalNonlinException('rms_upp_limit must be greater than rms_low_limit')
+
+    if not isinstance(lowess_frac, (float, int)):
+        raise TypeError('lowess_frac is not a number')
+    if not isinstance(rms_low_limit, (float, int)):
+        raise TypeError('rms_low_limit is not a number')
+    if not isinstance(rms_upp_limit, (float, int)):
+        raise TypeError('rms_upp_limit is not a number')
+    if not isinstance(pfit_upp_cutoff1, (float, int)):
+        raise TypeError('pfit_upp_cutoff1 is not a number')
+    if not isinstance(pfit_upp_cutoff2, (float, int)):
+        raise TypeError('pfit_upp_cutoff2 is not a number')
+    if not isinstance(pfit_low_cutoff1, (float, int)):
+        raise TypeError('pfit_low_cutoff1 is not a number')
+    if not isinstance(pfit_low_cutoff2, (float, int)):
+        raise TypeError('pfit_low_cutoff2 is not a number')
+
+    if make_plot is True:
+        # Avoid issues with importing matplotlib on headless servers without GUI
+        # support without proper configuration
+        import matplotlib.pyplot as plt
+        # Output directory
+        if os.path.exists(plot_outdir) is False:
+            os.mkdir(plot_outdir)
+            if verbose:
+                print('Output directory for figures created in ', os.getcwd())
+    
+    ######################### start of main code #############################
+    
+    # Define offset columns ROI range
+    offset_colroi = slice(offset_colroi1,offset_colroi2)
+    
+    # Define pixel ROIs
+    rowroi = list(range(rowroi1, rowroi2))
+    colroi = list(range(colroi1, colroi2))
+    
+    # Background subtraction regions
+    rowback1 = list(range(rowback11, rowback12))
+    rowback2 = list(range(rowback21, rowback22))
+    colback1 = list(range(colback11, colback12))
+    colback2 = list(range(colback21, colback22))
+    
+    ####################### create good_mean_frame ###################
+    
+    nrow = len(mean_frame_arr[0])
+    ncol = len(mean_frame_arr[0][0])
+    
+    good_mean_frame = np.zeros((nrow, ncol))
+    nFrames = len(mean_frame_arr)
+
+    good_mean_frame = good_mean_frame / nFrames
+    
+    mean_frame_index = 0
+    # Loop over the mean_frame_arr frames
+    for i in range(nFrames):
+        frame = mean_frame_arr[i]
+        frame = frameProc(frame.astype(np.float64), offset_colroi,
```

Comment:
same comment as in Kgain. frameProc has been superseded by `prescan_biassub` and `em_gain_division`. I think it would be fair to make it a requirement that the input dataset has already been processed by these steps.

Reply:
Waiting to hear about the changes done in calibrate_kgain in this respect

Reply:
Sure this looks fine.